File: server/utils/queue/blockListManager.js
import { getBlockListJson, saveBlockListJson, checkSimilarity } from "../utils.js";
import BaseQueueManager from "./baseQueueManager.js";
import logger from "../logger.js";

class BlockListManager extends BaseQueueManager {
    constructor() {
        super({
            readFunction: getBlockListJson,
            saveFunction: saveBlockListJson,
            validateFunction: (item) => {
                return typeof item === "object" && item.songName;
            },
            formatFunction: (item) => ({
                ...item,
                blockedAt: item.blockedAt || new Date().toISOString()
            })
        });
    }

    #isSimilarSong(songName1, songName2) {
        const similarity = checkSimilarity(songName1, songName2);
        return similarity >= 85;
    }

    async blockCurrentSong(songName, requestedBy) {
        try {
            if (!songName) {
                throw new Error("Song name is required");
            }

            if (this.isSongBlocked(songName)) {
                logger.warn(`Song already blocked: ${songName}`);
                return "Song is already in block list.";
            }

            const blockItem = {
                songName,
                requestedBy,
                blockedAt: new Date().toISOString()
            };

            this.add(blockItem);
            logger.info(`Blocked song: ${songName}`);
            return "Blocked the current song.";
        } catch (error) {
            logger.error("Error blocking current song:", {error});
            throw error;
        }
    }

    async blockSongBySongName(songName, requestedBy) {
        try {
            if (!songName) {
                throw new Error("Song name is required");
            }

            return await this.blockCurrentSong(songName, requestedBy);
        } catch (error) {
            logger.error("Error blocking song by name:", {error});
            throw error;
        }
    }

    async unblockSongBySongName(songName) {
        try {
            if (!songName) {
                throw new Error("Song name is required");
            }

            const index = this.items.findIndex(item => this.#isSimilarSong(item.songName, songName));
            if (index === -1) {
                logger.warn(`Song not found in block list: ${songName}`);
                throw new Error("Song not found in block list");
            }

            this.removeAtIndex(index + 1); // +1 because removeAtIndex expects 1-based index
            logger.info(`Unblocked song: ${songName}`);
            return "Unblock Successful";
        } catch (error) {
            logger.error("Error unblocking song by name:", { error });
            throw error;
        }
    }

    async unblockSongByIndex(index) {
        try {
            const result = this.removeAtIndex(index);
            if (result) {
                logger.info(`Unblocked song at index ${index}: ${result.songName}`);
                return "Unblock Successful.";
            }
            throw new Error("Invalid index");
        } catch (error) {
            logger.error("Error unblocking song by index:", {error});
            throw error;
        }
    }

    async clearBlockList() {
        try {
            this.clear();
            logger.info("Cleared block list");
            return "Cleared the block list.";
        } catch (error) {
            logger.error("Error clearing block list:", {error});
            throw error;
        }
    }

    async getAllBlockList() {
        try {
            return this.getAll();
        } catch (error) {
            logger.error("Error getting block list:", {error});
            throw error;
        }
    }

    isSongBlocked(songName) {
        return this.items.some(item => this.#isSimilarSong(item.songName, songName));
    }
}

export default BlockListManager;

File: app/managers/block_list_manager.py
from datetime import datetime
from app.managers.base_queue_manager import BaseQueueManager
from app.core.utils import get_block_list_json, save_block_list_json, check_similarity
from app.core import logger

class BlockListManager(BaseQueueManager):
    def __init__(self):
        def read_function():
            return get_block_list_json()
        
        def save_function(items):
            return save_block_list_json(items)
        
        def validate_function(item):
            return isinstance(item, dict) and item.get('songName')
        
        def format_function(item):
            return {
                **item,
                'blockedAt': item.get('blockedAt') or datetime.now().isoformat()
            }
        
        super().__init__({
            'read_function': read_function,
            'save_function': save_function,
            'validate_function': validate_function,
            'format_function': format_function
        })
    
    def _is_similar_song(self, song_name1, song_name2):
        similarity = check_similarity(song_name1, song_name2)
        return similarity >= 85
    
    async def block_current_song(self, song_name, requested_by):
        try:
            if not song_name:
                raise Exception("Song name is required")
            
            if self.is_song_blocked(song_name):
                logger.warn(f"Song already blocked: {song_name}")
                return "Song is already in block list."
            
            block_item = {
                'songName': song_name,
                'requestedBy': requested_by,
                'blockedAt': datetime.now().isoformat()
            }
            
            self.add(block_item)
            logger.info(f"Blocked song: {song_name}")
            return "Blocked the current song."
        except Exception as error:
            logger.error("Error blocking current song", error=str(error))
            raise error
    
    async def block_song_by_song_name(self, song_name, requested_by):
        try:
            if not song_name:
                raise Exception("Song name is required")
            
            return await self.block_current_song(song_name, requested_by)
        except Exception as error:
            logger.error("Error blocking song by name", error=str(error))
            raise error
    
    async def unblock_song_by_song_name(self, song_name):
        try:
            if not song_name:
                raise Exception("Song name is required")
            
            index = None
            for i, item in enumerate(self.items):
                if self._is_similar_song(item['songName'], song_name):
                    index = i
                    break
            
            if index is None:
                logger.warn(f"Song not found in block list: {song_name}")
                raise Exception("Song not found in block list")
            
            self.remove_at_index(index + 1)
            logger.info(f"Unblocked song: {song_name}")
            return "Unblock Successful"
        except Exception as error:
            logger.error("Error unblocking song by name", error=str(error))
            raise error
    
    async def unblock_song_by_index(self, index):
        try:
            result = self.remove_at_index(index)
            if result:
                logger.info(f"Unblocked song at index {index}: {result['songName']}")
                return "Unblock Successful."
            raise Exception("Invalid index")
        except Exception as error:
            logger.error("Error unblocking song by index", error=str(error))
            raise error
    
    async def clear_block_list(self):
        try:
            self.clear()
            logger.info("Cleared block list")
            return "Cleared the block list."
        except Exception as error:
            logger.error("Error clearing block list", error=str(error))
            raise error
    
    async def get_all_block_list(self):
        try:
            return self.get_all()
        except Exception as error:
            logger.error("Error getting block list", error=str(error))
            raise error
    
    def is_song_blocked(self, song_name):
        return any(self._is_similar_song(item['songName'], song_name) for item in self.items)
